lib/SILOptimizer/Transforms/OptRemarkGenerator.cpp

```diff
@@ -23,6 +24,82 @@
 
 using namespace swift;
 
+//===----------------------------------------------------------------------===//
+//                                  Utility
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+/// Given a value, call \p funcPassedInferredArgs for each associated
+/// ValueDecl that is associated with \p value. All created Arguments are
+/// passed the same StringRef. To stop iteration, return false in \p
+/// funcPassedInferedArgs.
+///
+/// NOTE: the function may be called multiple times if the optimizer joined
+/// two live ranges and thus when iterating over value's users we see multiple
+/// debug_value operations.
+struct ValueToDeclInferrer {
+  using Argument = OptRemark::Argument;
+  using ArgumentKeyKind = OptRemark::ArgumentKeyKind;
+
+  bool infer(ArgumentKeyKind keyKind, StringRef msg, SILValue value,
```

Comment:
Part of the reason why I am doing this is I am going to start storing state in here. Currently just projections that we strip off (which merge into the diagnostic emitted.

Reply:
I may also change the callback function to just be an out array as well.